keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -348,25 +349,168 @@ to fail with a descriptive error.
+- replace
+- sublist
+- split
+- reverse
+- sum/min/max (xref: [Rego function library](https://www.openpolicyagent.org/docs/latest/policy-reference/#built-in-functions))
+
+Numbers:
+
+- abs/cel/floor/round
+- exp/log/log10/logN/pow/sqrt
+- max(x, y, ..)/min(x, y, ...)
+- <trig functions>
+
+Formats:
+
+- semver: compare (xref [Kyverno function library](https://github.com/kyverno/kyverno/blob/main/pkg/engine/jmespath/functions.go))
+
+Maps:
+
+- <Ability to construct a map using a comprehension>
 
+Some considerations when selecting which of the above we should include in CEL expressions in Kubernetes:
+
+- We should include functions that will be needed for future use cases of CEL in Kubernetes (e.g. general admission) as
+  ones needed for validation.
+- Regex replace is very powerful and useful. It is also potentially dangerous due to its ability to allocate memory.
+- `format`: Since CEL supports string concatenation, the value of having format would only be to do things like format floats.
+  Formatting functions are complex and require extensive documentation to teach.
+- `isSorted` makes it possible to check if a list is sorted without the expense of performing a sort
+- `indexOf` / `lastIndexOf` / `split` / `replace` (the string functions) will be overloaded to provide the equivalent list functions.
+- `reduce` is going to be non-obvious to anyone that doesn't have a functional programming background?
+  - If we provide `reduce` are developers going to also expect to have `fold` or `zip`?
+- "Ability to construct a map using a comprehension": this appears mechanically problematic to support in CEL?
+- None of the other policy libraries provide extended math/trig support and I asked Tristan (maintainer of CEL) if
+  they are requested and he said "almost never", which he clarified to mean it has been requested exactly one time.
+- Average and quantile (median, 99th percentile, ...) and stddev are, however, often requested by other CEL users
+
+Future work:
+
+We've decided NOT to add the following functions. They may be useful for mutation, in which case we will consider adding them as part
+of any future work done involving CEL and mutating admission control:
+
+- Regex replace
+- Add `trimPrefix` / `trimSuffix` for strings
+- Add `trim` / `trimLeft` / `trimRight` (overloaded to take an optional cutset arg) for strings
+- `sublist`, `split`, `replace` and `reverse` for lists (overloading existing string functions where appropriate)
+
+##### Function Library Updates
+
+Any changes to the CEL function library will make it possible for CRD authors to create CRDs that are incompatible with
+all previous Kubernetes releases that supported CEL. Because of this incompatibility, changing the function library
+will need to carefully considered and kept to a minimum. All changes will need to be limited to function additions.
+We will not add functions to do things that can already be accomplished with existing functions. Improving ease-of-use
```

Comment:
I'd like to explore the limits on this.  Is it possible to take a URL (scheme://host:<port>) and parse out the host if we allow
1. dns names
2. ipv4
3. ipv6
4. ipv4 mapped in ipv6

I'm interested in how difficult this is to do with existing functions and if it is very difficult, would we really hold this line?  I see it messed up hand-written parsing code very frequently.

Reply:
I think it's intractable with what we have in CEL now due to the lack of regex `Find`.

Options:

1. Add regex find and findAll - [Very unlikely](https://community.helpsystems.com/forums/intermapper/miscellaneous-topics/5acc4fcf-fa83-e511-80cf-0050568460e4?_ga=2.113564423.1432958022.1523882681-2146416484.1523557976) that people will get the regex exactly right
2. Add URL parsing functions.  This should probably be: string -> URL conversion + getScheme, getUserinfo, getHost, getPath, getQuery, getFragment

I'm currently in favor of adding both. We made an explicit decision to not add regex replace (shouldn't really be needed for validation), but I hadn't thought carefully about match vs. find.